src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,144 @@
+    }
+
+    private void ensureCapacity() {
+        T[] newElements = (T[]) new Object[elements.length * 2];
+        copyArray(elements, 0, newElements, 0, size);
+        elements = newElements;
+    }
+
+    private void copyArray(T[] arrayFrom,
+                           int arrayFromPos,
+                           T[] arrayTo,
+                           int arrayToPos,
+                           int amount) {
+        for (int i = 0; i < amount; i++) {
+            arrayTo[arrayToPos + i] = arrayFrom[arrayFromPos + i];
+        }
+    }
+
+    private boolean isFull() {
+        return size == elements.length;
+    }
+
+    private boolean isIndexValid(int index) {
+        return index < size && index >= 0;
+    }
+
+    private boolean isIndexValidForAdd(int index) {
+        return index <= size && index >= 0;
+    }
+
     @Override
     public void add(T value) {
-
+        elements[size] = value;
+        size++;
+        if (isFull()) {
+            ensureCapacity();
+        }
     }
 
     @Override
     public void add(T value, int index) {
+        if (!isIndexValidForAdd(index)) {
+            throw new ArrayListIndexOutOfBoundsException(
+                    "You tried to add element to non exist index"
+            );
+        }
+
+        for (int i = size; i > index; i--) {
+            elements[i] = elements[i - 1];
+        }
+        elements[index] = value;
+        size++;
 
+        if (isFull()) {
+            ensureCapacity();
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            elements[size] = list.get(i);
+            size++;
+            if (isFull()) {
+                ensureCapacity();
+            }
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (!isIndexValid(index)) {
+            throw new ArrayListIndexOutOfBoundsException("You"
+                    + " tried to get element by non existed index");
+        }
+        return elements[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (!isIndexValid(index)) {
+            throw new ArrayListIndexOutOfBoundsException(
+                    "You tried to set element to non exist index"
+            );
+        }
+        elements[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (!isIndexValid(index)) {
+            throw new ArrayListIndexOutOfBoundsException(
+                    "You tried to remove element from non exist index"
+            );
+        }
+        final T elementToRemove = elements[index];
+        T[] newElements = (T[]) new Object[elements.length];
+        size--;
+        copyArray(elements, 0, newElements, 0, index);
+        copyArray(elements, index + 1, newElements, index, size - index);
+        elements = newElements;
+        return elementToRemove;
```

Comment:
There is no need to create a new array each time you remove an element. Instead, you can shift all elements to the left from the index of the removed element. Also, consider the case where the array is almost empty but its length is huge. It may be a good idea to decrease the size of the array in this case.